### PR TITLE
[CCXDEV-16284] Implement memleak workaround from insights-core

### PR DESCRIPTION
**Pull request by matysek**
# Description

[CCXDEV-15098] Implement memleak workaround from insights-core

Workaround for the insights-core fix from: https://github.com/RedHatInsights/insights-core/pull/4763

## Type of change

- Bug fix (non-breaking change which fixes an issue)

[CCXDEV-15098]: https://redhat.atlassian.net/browse/CCXDEV-15098?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: .github/workflows/pytests.yaml [hunk 1]
```diff
@@ -17,6 +17,7 @@ jobs:
         python-version:
           - "3.11"
           - "3.12"
+          - "3.13"
```
Comment (ikerreyes) [inline .github/workflows/pytests.yaml]: how is this related?

File: test/test_traceback_leak.py [hunk 2]
```diff
@@ -0,0 +1,271 @@
...
+def _make_exception_with_traceback(exc_type=Exception, msg="test exception"):
+    """Raise and catch an exception so it has a real __traceback__."""
+    try:
+        raise exc_type(msg)
+    except exc_type as ex:
+        tb_string = traceback.format_exc()
+        return ex, tb_string
+
+
+def test_traceback_cleared_keyboard_interrupt():
```
Comment (ikerreyes) [inline test/test_traceback_leak.py:23]: Why are we concerned about keyboard interrupts for a service?

File: test/test_traceback_leak.py [hunk 3]
```diff
@@ -0,0 +1,271 @@
...
+def test_sentry_not_called_for_excluded_types():
+    """MissingRequirements and ContentException skip Sentry but still clear traceback."""
+    broker = SentryMonitoredBroker()
+
+    for exc_type, msg in [
+        (MissingRequirements, ["req"]),
+        (ContentException, "content"),
+    ]:
+        ex = exc_type(msg)
+        try:
```
Comment (ikerreyes) [inline test/test_traceback_leak.py:93]: why not using `_make_exception_with_traceback` here?

File: test/test_traceback_leak.py [hunk 4]
```diff
@@ -0,0 +1,271 @@
...
+    ex = Exception("never raised")
+    assert ex.__traceback__ is None
+
+    with patch("ccx_messaging.monitored_broker.capture_exception"):
+        broker.add_exception("component", ex, None)
+
+    assert ex.__traceback__ is None
+
+
+def test_broker_collected_after_add_exception():
```
Comment (ikerreyes) [inline test/test_traceback_leak.py:219]: I am not understanding the intent of this test. Can you explain?

**Comment by matysek** [comment]
@coderabbitai review

File: .github/workflows/pytests.yaml (hunk 1 above)
Comment (matysek) [reply .github/workflows/pytests.yaml]: you are right, reverted.

File: test/test_traceback_leak.py (hunk 3 above)
Comment (matysek) [reply test/test_traceback_leak.py:93]: Updated docstring - make it explicit that exceptions have different constructor types.

File: test/test_traceback_leak.py (hunk 4 above)
Comment (matysek) [reply test/test_traceback_leak.py:219]: Added comment better explaining the test.

File: test/test_traceback_leak.py (hunk 2 above)
Comment (matysek) [reply test/test_traceback_leak.py:23]: KeyboardInterrupt exception is in some frameworks mapped to signal SIGTERM for graceful shutdown. Also updated the doscring with better explanation.

File: ccx_messaging/monitored_broker.py [hunk 5]
```diff
@@ -22,4 +22,37 @@ def add_exception(self, component, ex, tb=None):
...
+        # TODO: remove when this is merged: https://github.com/RedHatInsights/insights-core/pull/4763
+        # Python attaches a live traceback to ex.__traceback__ which
+        # references the stack frame (including the broker), creating:
+        #   broker.exceptions -> ex -> __traceback__ -> frame -> broker
+        # The formatted traceback string is already stored in
+        # broker.tracebacks via super().add_exception(), and
+        # capture_exception() above receives the live traceback for
+        # error-tracking systems (Sentry/GlitchTip), so no debugging
+        # information is lost.
+        if isinstance(ex, BaseException):
```
Comment (ikerreyes) [inline ccx_messaging/monitored_broker.py:42]: Is there a case where `ex` is not expected to derive from BaseException?

File: test/test_traceback_leak.py [hunk 6]
```diff
@@ -0,0 +1,299 @@
...
+            ex, tb = _make_exception_with_traceback(Exception, "gc test")
+
+            with patch("ccx_messaging.monitored_broker.capture_exception"):
+                broker.add_exception("component", ex, tb)
+
+            refs.append(weakref.ref(broker))
+            del broker, ex, tb
+
+        collected = sum(1 for ref in refs if ref() is None)
+        assert collected >= n_brokers - 1, (
```
Comment (ikerreyes) [inline test/test_traceback_leak.py]: Why allowing for 1 broker to not match the criteria?

File: pyproject.toml [hunk 7]
```diff
@@ -28,7 +28,7 @@ dependencies = [
     "boto3>=1.42.91,<1.42.92",  # limited due to incompatibility between aiobotocore and newer versions of boto
     "confluent-kafka>=2.0.0",
     "insights-core>=3.1.2",
-    "insights-core-messaging>=1.2.15",
+    "insights-core-messaging @ git+https://github.com/RedHatInsights/insights-core-messaging@1.2.21",
```
Comment (ikerreyes) [inline pyproject.toml:31]: Cannot we release insights-core-messaging and use the PyPI package?

Comment (matysek) [reply pyproject.toml:31]: We do not  have permissions yet for the pypi package.

File: ccx_messaging/monitored_broker.py (hunk 5 above)
Comment (matysek) [reply ccx_messaging/monitored_broker.py:42]: It is a defensive check. Added comment. I agree that it is unlikely in prod env.

File: test/test_traceback_leak.py (hunk 6 above)
Comment (matysek) [reply test/test_traceback_leak.py]: Added comment. That is there for tests to pass  even for edge cases.

Comment (ikerreyes) [reply test/test_traceback_leak.py]: But won't that also prevent finding some mismatch/wrong config?

Comment (matysek) [reply test/test_traceback_leak.py]: Ok, let's rather thne try the `collected >= n_brokers`. If there are later any issues with test stability, we will see.

**Review by ikerreyes — APPROVED** [review]

**Review by JiriPapousek — APPROVED** [review]

**Review by Bee-lee — APPROVED** [review]
